File: Projects/PyWebDev/validation.py
import re
import extensions

# route can be user, user_edit or login
# return a list of error messages
# code: 1. username is taken
#       2. username at least 3 characters long
#       3. username only contains letters, digits, underscores
#       4. password at least 8 characters long
#       5. password contain at least one letter and one number
#       6. password only contain letters, digits, underscores
#       7. password not match
#       8. email not valid
#       9. username max length 20
#       10. firstname max length 20
#       11. lastname max length 20
#       12. email max length 40
#       13. username does not exist
#       14. password incorrect

def check_errors(route, user_info):
    error_dict = {};
    if "username" in user_info:
        if (route == "user"):
            if (len(user_info["username"]) > 20):
                error_dict['9'] = 1;
            if (len(user_info["username"]) < 3):
                error_dict['2'] = 1;
            if (re.match(r"^[\w]*$", user_info["username"]) is None):
                error_dict['3'] = 1;
            if (extensions.does_username_exist(user_info["username"])):
                error_dict['1'] = 1;

    if "password1" in user_info:
        if user_info['password1'] == "" and user_info['password2'] == "" and route == 'user_edit':
	    pass
        elif (route == "user" or route == "user_edit"):
            if (len(user_info["password1"]) < 8):
		error_dict['4'] = 1;
            if (re.match(r"^[\w]*$", user_info["password1"]) is None):
                error_dict['6'] = 1;
            if ((re.match(r"^.*[a-zA-Z].*\d.*$", user_info["password1"]) is None) and (re.match(r"^.*\d.*[a-zA-Z].*$", user_info["password1"]) is None)):
                error_dict['5'] = 1;
            if (user_info["password1"] != user_info["password2"]):
                error_dict['7'] = 1;

    if "email" in user_info:
        if (route == "user" or route == 'user_edit'):
            if (len(user_info["email"]) > 40):
                error_dict['12'] = 1;
            if not re.match(r"[^@]+@[^@]+\.[^@]+", user_info["email"]):
                error_dict['8'] = 1;

    if "firstname" in user_info:
        if (route == "user" or route == 'user_edit'):
            if (len(user_info["firstname"]) > 20):
                error_dict['10'] = 1;

    if "lastname" in user_info:
        if (route == "user" or route == 'user_edit'):
            if (len(user_info["lastname"]) > 20):
                error_dict['11'] = 1;

    if (route == "login"):
        if (not extensions.does_username_exist(user_info["username"])):
            error_dict['13'] = 1;
        else:
            if not extensions.check_login(username=user_info["username"], password=user_info["password"]):
                error_dict['14'] = 1;

    return error_dict;

File: Projects/PyWebDev/static/js/validation.js

function check_errors(route, user_info) {
    var error_msg_list = [];

    if ("username" in user_info) {
        if (route == "user" || route == "user_edit") {

            if (user_info["username"].length > 20) {
                error_msg_list.push("Username must be no longer than 20 characters");
            }

            if (user_info["username"].length < 3) {
                error_msg_list.push("Usernames must be at least 3 characters long");
            }

            if (user_info["username"].search(/^[\w]*$/) == -1) {
                error_msg_list.push("Usernames may only contain letters, digits, and underscores");
            }
        }
    }



    if ("password1" in user_info) {
        if (route == "user_edit" && user_info["password1"] == "" && user_info["password2"] == "") {
        }

        else if (route == "user" || route == "user_edit") {
            if (user_info["password1"].length < 8) {
                error_msg_list.push("Passwords must be at least 8 characters long");
            }
            if (user_info["password1"].search(/^[\w]*$/) == -1) {
                error_msg_list.push("Passwords may only contain letters, digits, and underscores");
            }

            if ((user_info["password1"].search(/^.*[a-z].*\d.*$/i) == -1) && (user_info["password1"].search(/^.*\d.*[a-z].*$/i) == -1)) {
                error_msg_list.push("Passwords must contain at least one letter and one number");
            }

            if (user_info["password1"] != user_info["password2"]) {
                error_msg_list.push("Passwords do not match");
            }
        }
    }



    if ("email" in user_info) {
        if (route == "user" || route == "user_edit") {
            if (user_info["email"].length > 40) {
                error_msg_list.push("Email must be no longer than 40 characters");
            }

            if (user_info["email"].search(/[^@]+@[^@]+\.[^@]+/) == -1) {
                error_msg_list.push("Email address must be valid");
            }
        }
    }


    if ("firstname" in user_info) {
        if (route == "user" || route == "user_edit") {
            if (user_info["firstname"].length > 20) {
                error_msg_list.push("Firstname must be no longer than 20 characters");
            }
        }
    }


    if ("lastname" in user_info) {
        if (route == "user" || route == "user_edit") {
            if (user_info["lastname"].length > 20) {
                error_msg_list.append("Lastname must be no longer than 20 characters");
            }
        }
    }

    return error_msg_list;
}


